nautobot/dcim/models/device_components.py

```diff
@@ -1096,3 +1154,101 @@ def natural_key_field_lookups(cls):  # pylint: disable=no-self-argument
         For the time being we just use the PK as a natural key.
         """
         return ["pk"]
+
+
+@extras_features(
+    "custom_links",
+    "custom_validators",
+    "export_templates",
+    "graphql",
+    "webhooks",
+)
+class ModuleBay(PrimaryModel):
+    """
+    A slot in a Device or Module which can contain Modules.
+    """
+
+    parent_device = models.ForeignKey(
+        to="dcim.Device",
+        on_delete=models.CASCADE,
+        related_name="module_bays",
+        blank=True,
+        null=True,
+    )
+    parent_module = models.ForeignKey(
+        to="dcim.Module",
+        on_delete=models.CASCADE,
+        related_name="module_bays",
+        blank=True,
+        null=True,
+    )
+    position = models.CharField(
+        blank=False,
+        null=False,
+        max_length=CHARFIELD_MAX_LENGTH,
+        help_text="The position of the module bay within the parent device/module",
+    )
+    label = models.CharField(max_length=CHARFIELD_MAX_LENGTH, blank=True, help_text="Physical label")
+    description = models.CharField(max_length=CHARFIELD_MAX_LENGTH, blank=True)
+
+    clone_fields = ["parent_device", "parent_module"]
+
+    class Meta:
+        # TODO: Ordering by parent_module.id is not correct but prevents an infinite loop
+        ordering = (
+            "parent_device",
+            "parent_module__id",
+            "position",
+        )
+        constraints = [
+            models.CheckConstraint(
+                check=models.Q(parent_device__isnull=False, parent_module__isnull=True)
+                | models.Q(parent_device__isnull=True, parent_module__isnull=False),
+                name="dcim_modulebay_parent_device_xor_parent_module",
+            ),
+            models.UniqueConstraint(
+                fields=["parent_device", "position"],
+                name="dcim_modulebay_parent_device_position_unique",
+            ),
+            models.UniqueConstraint(
+                fields=["parent_module", "position"],
+                name="dcim_modulebay_parent_module_position_unique",
+            ),
+        ]
+
+    @property
+    def parent(self):
+        return self.parent_device if self.parent_device else self.parent_module
```

Comment:
Would this work since they are guaranteed to be mutually exclusive?
```suggestion
        return self.parent_device | self.parent_module
```